### PR TITLE
test(locality-ratchet): adversarial gate (qa-d, #201)

### LINKED ISSUE #201: QA gate: ratchet enforcement composition adversarial test (qa-d)
## What to build

Adversarial QA gate validating ratchet enforcement composition. Three hooks (pre-commit blocker, rolling-window budget, PostToolUse advisory) compose across pre-commit + post-edit boundaries. This gate attacks the composed surface with malicious commit sequences and edge cases.

## Acceptance criteria

- [ ] **Compositional fixtures** (commit sequences exercising both hooks):
  - [ ] Commit with trailer in commit body (not subject) — both hooks parse it correctly
  - [ ] Commit with multiple trailers in one message — counted as one (not double-credit)
  - [ ] Commit that touches both exempted region (matrix table) AND non-exempted region in a single hunk — verify hook handles partial allow / partial block correctly
- [ ] **Rolling-window edge cases:**
  - [ ] Window wraparound: commits at the exact window boundary (timestamp-based)
  - [ ] Concurrent commits (test parallel pre-commit invocations on the same repo state)
  - [ ] Paired-deletion in same commit as new addition (does it count as both a pairing AND a use of the budget? Pick semantics, test it)
- [ ] **Bypass attempts:**
  - [ ] Trailer with empty value
  - [ ] Trailer in unsigned commit (verify hook doesn't depend on signature)
  - [ ] Amend to add trailer after the fact (`git commit --amend`)
  - [ ] `--no-verify` bypass — hook obviously can't block this, but document the bypass + mitigation (CI re-runs hook)
- [ ] **Cross-hook race:** PostToolUse advisory fires during pre-commit hook execution — verify no deadlock, no interleaved output corruption
- [ ] All findings logged; cross-functional review chain completes; P0–P2 remediated before close

## Blocked by

- #199 (slice 5a basic enforcement)
- #200 (slice 5b rolling-window budget)
- #195 (PostToolUse advisory)

### PR DESCRIPTION
**Pull request by wryenmeek**
> *Implemented by AI fleet agent. Closes #201.*

## Summary
- Add adversarial locality-ratchet tests for trailer parsing, non-trailer bypass attempts, AGENTS matrix/non-exempt composition, Locality 0 invariant interaction, and concurrent precommit/advisory execution.
- Mark known content-insensitive paired-deletion cases as strict xfails and track the production-hook fix in #266.
- Update the audit-workspace improve-flow progress table for qa-d.

## Tests
- `python3 -m pytest tests/kb/test_locality_ratchet_adversarial.py tests/kb/test_locality_ratchet.py -v` — 25 passed, 2 xfailed

### CODE REVIEW

File: tests/kb/test_locality_ratchet_adversarial.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+    return subprocess.run(
+        command,
+        cwd=repo,
+        env=env,
+        input=stdin_text,
+        capture_output=True,
+        text=True,
+        check=False,
+        timeout=10,
+    )
```
[the remark this answers is not in the document]
Comment (wryenmeek) [reply tests/kb/test_locality_ratchet_adversarial.py:138]: > Explicitly passing an empty string when `stdin_text` is `None` ensures that stdin is always redirected and isolated.

Addressed: `_run_ratchet_hook` now passes `input=stdin_text if stdin_text is not None else ""`, so subprocess stdin is always isolated. Re-ran `python3 -m pytest tests/kb/test_locality_ratchet_adversarial.py tests/kb/test_locality_ratchet.py -v` — 25 passed, 2 xfailed.
